### PR TITLE
Make mapping a distinct concept in logsdb data generation

### PR DESCRIPTION
**Pull request by lkts**
This PR splits data generation logic into mapping generation and document generation. This split allows to change the resulting mapping/document after it was generation or tweak mapping generation separately from document generation on a per-test basis. It also (arguably) simplifies the data generator.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingGenerator.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+            // copy_to can't be used across nested documents so all currently eligible fields are not eligible inside nested document.
+            return new Context(nested ? new HashSet<>() : eligibleCopyToDestinations, pathTo(name), subobjects, dynamicMapping);
+        }
+
+        void addCopyToCandidate(String leafFieldName) {
+            eligibleCopyToDestinations.add(pathTo(leafFieldName));
+        }
+
+        String pathTo(String leafFieldName) {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingGenerator.java:162]: Nit: dedup

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingTemplate.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+package org.elasticsearch.logsdb.datageneration;
+
+import java.util.Map;
+
+public record MappingTemplate(Map<String, Entry> mapping) {
+    public sealed interface Entry permits Entry.Leaf, Entry.Object {
+        record Leaf(String name, FieldType type) implements Entry {}
+
+        record Object(String name, boolean nested, Map<String, Entry> children) implements Entry {}
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingTemplate.java:18]: Nit: I'd define these outside the interface to keep the name shorter.. Still within MappingTemplate so no confusion.

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingTemplate.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.logsdb.datageneration;
+
+import java.util.Map;
+
+public record MappingTemplate(Map<String, Entry> mapping) {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingTemplate.java:14]: Nit: let's add comment classes everywhere, so it's easier for someone to understand what functionality is provided.

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/fields/leaf/ByteFieldDataGenerator.java [hunk 4]
```diff
@@ -54,4 +54,9 @@ public CheckedConsumer<XContentBuilder, IOException> mappingWriter() {
     public CheckedConsumer<XContentBuilder, IOException> fieldValueGenerator() {
         return b -> b.value(valueGenerator.get());
     }
+
+    @Override
+    public Object generateValue() {
+        return valueGenerator.get();
+    }
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/fields/leaf/ByteFieldDataGenerator.java:61]: These three functions seem to repeat.. maybe just define them in `FieldDataGenerator` and pass the field type name.

File: test/framework/src/test/java/org/elasticsearch/logsdb/datageneration/MappingGeneratorTests.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        var templateGenerator = new MappingTemplateGenerator(specification);
+        var template = templateGenerator.generate();
+
+        var generator = new MappingGenerator(specification);
+        var mapping = generator.generate(template, null);
+        validateMapping(mapping.raw());
+    }
+
+    private void validateMapping(Map<String, Object> mapping) {
+
```
Comment (kkrik-es) [inline test/framework/src/test/java/org/elasticsearch/logsdb/datageneration/MappingGeneratorTests.java:29]: Add a TODO?

**Comment by kkrik-es** [comment]
The general direction seems fine to me.. I like that mappings follow the same structure as in `MappingLookup`,  they'll make extending the generator and debugging tests easier.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/DataGenerationHelper.java [hunk 6]
```diff
@@ -62,45 +77,83 @@ public CheckedConsumer<XContentBuilder, IOException> mappingWriter() {
...
-        dataGenerator.writeMapping(builder);
+        builder.map(mapping.raw());
     }
 
     void standardMapping(XContentBuilder builder) throws IOException {
-        dataGenerator.writeMapping(builder);
+        var rawMapping = new HashMap<>(mapping.raw());
+        // We don't want standard to use custom subobjects values, this is not the goal of this test.
+        // We want to compare "clean" standard with logsdb.
+        removeSubobjects(rawMapping);
```
[the remark this answers is not in the document]
Comment (lkts) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/DataGenerationHelper.java:132]: I mostly have this as an example of how to use this new code.

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/DocumentGenerator.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+import org.elasticsearch.logsdb.datageneration.datasource.DataSourceRequest;
+import org.elasticsearch.logsdb.datageneration.datasource.DataSourceResponse;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+
+public class DocumentGenerator {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/DocumentGenerator.java:20]: Class comment please.

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/DocumentGenerator.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                } else {
+                    document.put(object.name(), generateObject(object, context));
+                }
+            }
+        }
+    }
+
+    private Map<String, Object> generateObject(Template.Object object, Context context) {
+        var children = new TreeMap<String, Object>();
+        generateDocument(children, object.children(), context.stepIntoObject(object.name()));
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/DocumentGenerator.java:76]: Nit: it's somewhat confusing that `generateObject` calls `generateDocument`.. Maybe rename the latter `generateContents` or `generateFields`

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingGenerator.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+                mappingParameters.putAll(mappingParametersGenerator.get());
+
+            } else if (templateEntry instanceof Template.Object object) {
+                boolean isDynamic = context.parentDynamicMapping != DynamicMapping.FORBIDDEN
+                    && dynamicMappingGenerator.generator().apply(true);
+                // Simply skip this field if it is dynamic.
+                // Lookup will contain null signaling dynamic mapping as well.
+                if (isDynamic) {
+                    continue;
+                }
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/MappingGenerator.java:122]: This is repeated, move before the branch?

**Review by kkrik-es — APPROVED** [review]
Looks good overall.. Let's make sure there are no new flaky failures due to some change in generation logic.

There are some pending comments, nothing major. Take a look and see what needs addressing.

**Comment by lkts** [comment]
@elasticmachine update branch

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/StandardVersusLogsIndexModeChallengeRestIT.java [hunk 10]
```diff
@@ -53,7 +53,7 @@
  * This test uses simple mapping and document structure in order to allow easier debugging of the test itself.
  */
 public class StandardVersusLogsIndexModeChallengeRestIT extends AbstractChallengeRestTest {
-    private final int numShards = randomBoolean() ? randomIntBetween(2, 5) : 0;
+    private final int numShards = randomBoolean() ? randomIntBetween(2, 4) : 0;
```
Comment (lkts) [inline modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/StandardVersusLogsIndexModeChallengeRestIT.java:56]: Without this tweak test cluster sometimes goes out of memory due to the size of translog. I assume this is just due to large documents.

**Pull request merged**